From b11aab13430b62775090ceb0cfccbc6da402c076 Mon Sep 17 00:00:00 2001
From: WofWca <wofwca@protonmail.com>
Date: Tue, 18 Nov 2025 12:34:11 +0400
Subject: [PATCH] fix: corpses don't absorb shotgun pellets

This is a follow-up to
ca9183fc5efa21b536fe11313079bf31f0e50e49
("fix: shotgun not gibbing unless aiming at feet")
(https://github.com/ec-/baseq3a/pull/49).
This brings the game balance closer to the original game.
For discussions, see
https://github.com/OpenArena/gamecode/pull/349.
---
 code/game/g_combat.c |  9 ++++-----
 code/game/g_weapon.c | 21 +++++++++++++++++++++
 2 files changed, 25 insertions(+), 5 deletions(-)

diff --git a/code/game/g_combat.c b/code/game/g_combat.c
index e29853ec..1e18a41f 100644
--- a/code/game/g_combat.c
+++ b/code/game/g_combat.c
@@ -607,11 +607,10 @@ void player_die( gentity_t *self, gentity_t *inflictor, gentity_t *attacker, int
 	// unless you aim at the feet.
 	// See https://github.com/ioquake/ioq3/issues/794.
 	//
-	// Note that without this line, when shooting at two players standing
-	// behind each other, the second target will take less damage,
-	// because the dead body of the first player will absorb the pellets
-	// until it gets gibbed (that is, up to 4 pellets,
-	// see `GIB_HEALTH` and `DEFAULT_SHOTGUN_DAMAGE`).
+	// Not executing this line makes is so that the corpse
+	// doesn't get shorter immediately on death
+	// and instead can still take up other pellets
+	// from the same shotgun shot.
 	//
 	// The purpose and the effect of this line is not entirely clear.
 	// Maybe it's to transition the player hitbox
diff --git a/code/game/g_weapon.c b/code/game/g_weapon.c
index d78afb0f..ea052cf0 100644
--- a/code/game/g_weapon.c
+++ b/code/game/g_weapon.c
@@ -306,6 +306,27 @@ qboolean ShotgunPellet( vec3_t start, vec3_t end, gentity_t *ent ) {
 				continue;
 			}
 #endif
+
+			// The below piece of code has been added in
+			// https://github.com/ec-/baseq3a/pull/60.
+			// When shooting through a corpse, gib it,
+			// but don't "absorb" the pellet, i.e. allow to hit a player
+			// through a corpse.
+			// This is mostly to compensate for the balance changes
+			// that are introduced by the removal of the `self->r.maxs[2] = -8;`
+			// line in `player_die`.
+			// But it's probably also sensible otherwise that corpses
+			// affect "more serious" gameplay less.
+			// See
+			// - https://github.com/ioquake/ioq3/issues/794
+			// - https://github.com/OpenArena/gamecode/pull/349
+			if ( traceEnt->client && traceEnt->client->ps.pm_type == PM_DEAD ) {
+				G_Damage( traceEnt, ent, ent, forward, tr.endpos, damage, 0, MOD_SHOTGUN );
+				passent = traceEnt->s.number;
+				VectorCopy( tr.endpos, tr_start );
+				continue;
+			}
+
 			if( LogAccuracyHit( traceEnt, ent ) ) {
 				hitClient = qtrue;
 			}
-- 
2.54.0

